### PR TITLE
Add `BooleanArray::new_from_packed` and `BooleanArray::new_from_u8`

### PR DESCRIPTION
**Pull request by chloro-pn**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 Users can easily build `BooleanArray` from `&[u8]`

# What changes are included in this PR?

* add `BooleanArray::new_from_packed`
* add `BooleanArray::new_from_u8`

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
I'm not sure about this, as it doesn't provide a mechanism to indicate a length that isn't a multiple of 8 or aligned.

Perhaps you could expand on what the use-case for this is, and why you would be using this as opposed to say `BooleanBufferBuilder` or similar?

**Comment by chloro-pn** [comment]
> I'm not sure about this, as it doesn't provide a mechanism to indicate a length that isn't a multiple of 8 or aligned.
> 
> Perhaps you could expand on what the use-case for this is, and why you would be using this as opposed to say `BooleanBufferBuilder` or similar?

Here, we only convert each bit in `&[u8]` to boolean and use it to build `BooleanArray`. It does not involve directly setting certain internal components, so there is no need to consider alignment and whether the length is a multiple of 8.
For example, in some scenarios, we need to use a bitstream (represented as `&[u8]` in Rust) to construct objects of type `BooleanArray`, where each bit in the bitstream is 1 for `true` and 0 for `false`.
If we didn't have this PR, we would always need to:
* first build `Buffer` from `&[u8]`, 
* then build `BooleanBuffer` from Buffer, and 
* finally build `BooleanArray` from `BooleanBuffer`.

Alternatively, we need to convert the bitstream into an iterator object that returns a boolean, but these come at a performance loss.

**Comment by chloro-pn** [comment]
Of course, using this method involves two self-evident facts:
* There is no `null` in the constructed `BooleanArray`
* The length of the constructed `BooleanArray` is always a multiple of 8;

If these prerequisites are not met, then this method should not be used.
But these will not cause confusion or difficulty for users to understand, as the mapping from bit stream to boolean array is intuitive.

**Comment by alamb** [comment]
I wonder if this related to the `8 bit boolean` extension type discussion on the dev list: https://lists.apache.org/thread/nz44qllq53h6kjl3rhy0531n2n2tpfr0 ?

File: arrow-array/src/array/boolean_array.rs [hunk 1]
```diff
@@ -349,6 +349,12 @@ impl From<Vec<Option<bool>>> for BooleanArray {
     }
 }
 
+impl From<&[u8]> for BooleanArray {
```
Comment (alamb) [inline arrow-array/src/array/boolean_array.rs:352]: If we do choose to add this feature I think it needs some additional docuemntation (e.g. an example on `BooleanArray` that shows how to use it and explains that it treats each byte as a 8 bool values, and that the resulting boolean array has length `bytes * 8` etc

**Review by alamb — COMMENTED** [review]
Thanks for this proposal @chloro-pn

**Comment by chloro-pn** [comment]
> I wonder if this related to the `8 bit boolean` extension type discussion on the dev list: https://lists.apache.org/thread/nz44qllq53h6kjl3rhy0531n2n2tpfr0 ?

No, we are not building a new type here.
We are just providing a more convenient way to build `BooleanArray` from bitstream, which will have better performance and be more convenient under the premise of meeting usage requirements.

File: arrow-array/src/array/boolean_array.rs (hunk 1 above)
Comment (chloro-pn) [reply arrow-array/src/array/boolean_array.rs:352]: > If we do choose to add this feature I think it needs some additional docuemntation (e.g. an example on `BooleanArray` that shows how to use it and explains that it treats each byte as a 8 bool values, and that the resulting boolean array has length `bytes * 8` etc

Okay, if this feature is accepted, I will provide some documentation: )

File: arrow-array/src/array/boolean_array.rs [hunk 2]
```diff
@@ -50,6 +50,22 @@ use std::sync::Arc;
 /// assert_eq!(&values, &[Some(true), None, Some(false), None, Some(false)])
 /// ```
 ///
+/// # Example: From `&[u8]`
```
Comment (alamb) [inline arrow-array/src/array/boolean_array.rs:53]: Thank you.

**Review by alamb — COMMENTED** [review]
I thought about this more last night. 

I think my main concern is if the semantics of `BooleanArray::from(&[u8])` will be "surprising" to people 

Thus I would like to propose making this a function like `BooleanArray::from_u8` like

```rust
let arr = BooleanArray::from_u8(..)
```

We can put the documentation on that method. I think that would avoid any concern about confusion (as users would could the method explicitly).

What do you think?

**Comment by chloro-pn** [comment]
> I thought about this more last night.
> 
> I think my main concern is if the semantics of `BooleanArray::from(&[u8])` will be "surprising" to people
> 
> Thus I would like to propose making this a function like `BooleanArray::from_u8` like
> 
> ```rust
> let arr = BooleanArray::from_u8(..)
> ```
> 
> We can put the documentation on that method. I think that would avoid any concern about confusion (as users would could the method explicitly).
> 
> What do you think?

I think your suggestion is more appropriate, and I will modify the code later.
The use of `&[u8]` to represent a bitstream is not so obvious and consistent (`&[u8]` can only represent a partial bitstream). 
The `From` trait is not only used to represent conversions between types, but also declares a semantically equivalent relationship.I get it.
I am a beginner in Rust, thanks for your code review: )

**Comment by tustvold** [comment]
Might I suggest something along the lines of 

```
fn from_packed(buffer: impl Into<Buffer>, offset: usize, len:usize)
```

This would be explicit and consistent with other methods

**Comment by chloro-pn** [comment]
> Might I suggest something along the lines of
> 
> ```
> fn new_from_packed(buffer: impl Into<Buffer>, offset: usize, len:usize)
> ```
> 
> This would be explicit and consistent with other methods.
> 
> Edit: although I wonder if this is then really all that different from
> 
> ```
> let a: BooleanArray = BooleanBuffer(b.into(), 0, len).into();
> ```
> 
> I wonder if we really need a new method...

This is what I mentioned in my previous reply "**providing a more convenient way to build BooleanArray from bitstream**".
As a developer who is very familiar with `arrow-rs`, this method is unnecessary, but for users, I think the provided interface should try to hide the internal structure as much as possible and maintain simplicity.

**Comment by tustvold** [comment]
My feeling is this convenience comes at the expense of being rather confusing to users, especially those less familiar with the internal layout of arrow arrays. My 2 cents is if people want a convenient and easy interface they can use the builders or iterators, if they want lower level integration they use lower level interfaces, providing a method that is convenient but correct usage requires knowledge of the lower level structure seems undesirable.

If Andrew is happy to move forward with this I have no objection to this, but I'm not sold

**Comment by alamb** [comment]
I agree that it takes a lot of knowledge until this is obvious:

```
let a: BooleanArray = BooleanBuffer(b.into(), 0, len).into();
```

It seems to me adding an example of how to do this in the documentation is valuable in any case.  

I also think it is valuable to add APIs that make arrow-rs easier to use for less expert users, even if that means they aren't the most concise, and thus I think an API such as proposd in the PR is a good addition

File: arrow-array/src/array/boolean_array.rs [hunk 3]
```diff
@@ -110,6 +110,24 @@ impl BooleanArray {
...
+    /// using this method will make the following points self-evident:
+    /// * there is no `null` in the constructed [`BooleanArray`];
+    /// * without considering `buffer.into()`, this method is efficient because there is no need to perform pack and unpack operations on boolean;
+    pub fn new_from_packed(buffer: impl Into<Buffer>, offset: usize, len: usize) -> Self {
+        BooleanBuffer::new(buffer.into(), offset, len).into()
+    }
+
+    /// Create a new [`BooleanArray`] from `&[u8]`
+    /// This method uses `new_from_packed` and constructs a [`Buffer`] using `value`, and offset is set to 0 and len is set to `value.len() * 8`
+    /// using this method will make the following points self-evident:
```
Comment (alamb) [inline arrow-array/src/array/boolean_array.rs:124]: ❤️

**Review by alamb — APPROVED** [review]
Thank you @chloro-pn  -- I think this is a nice addition to the API that is well commented and explains what it does well.

**Comment by alamb** [comment]
Thank you for the review and comments @tustvold

File: arrow-array/src/array/boolean_array.rs (hunk 3 above)
Comment (chloro-pn) [reply arrow-array/src/array/boolean_array.rs:124]: I think many people have said it before, but I still want to say that you are a very nice person. Thank you : )

**Comment by alamb** [comment]
Thanks again @chloro-pn  and @tustvold

**Pull request merged**